src/base/stun.ts

```diff
@@ -41,175 +42,322 @@ export const DEFAULT_PARALLEL_STUN_CALLS = 4
-      if (attr == null) {
-        error(`STUN response seems to have neither MappedAddress nor XORMappedAddress set. Dropping message`)
-        return
-      }
+        const attr = res.getXorMappedAddressAttribute() ?? res.getMappedAddressAttribute()
 
-      verbose(`Received STUN response. External address seems to be: ${attr.address}:${attr.port}`)
+        console.log = consoleBackup
 
-      if (results.push(attr) == usableMultiaddrs.length) {
-        done()
-      }
-    }
+        if (attr == null) {
+          error(`STUN response seems to have neither MappedAddress nor XORMappedAddress set. Dropping message`)
+          return
+        }
 
-    socket.on('message', msgHandler)
+        verbose(`Received STUN response. External address seems to be: ${attr.address}:${attr.port}`)
 
-    const allSent = performStunRequests(usableMultiaddrs, tids, socket)
+        if (addrs[index].response != undefined) {
+          verbose(`Recieved duplicate response. Dropping message`)
+          return
+        }
 
-    if (allSent.length > 0) {
-      const sendResults = await Promise.all(allSent)
+        addrs[index].response = attr
+        responsesReceived++
 
-      if (!sendResults.some((resultOk: boolean) => resultOk)) {
-        reject(
-          new Error(
-            `Cannot send any STUN packets. Tried with: ${usableMultiaddrs
-              .map((ma: Multiaddr) => ma.toString())
-              .join(', ')}`
-          )
-        )
+        if (responsesReceived == addrs.length) {
+          done()
+        }
+      } catch (err) {
+        consoleBackup(err)
+      } finally {
+        console.log = consoleBackup
       }
-
-      timeout = setTimeout(() => done(), STUN_TIMEOUT)
-    } else {
-      reject(new Error(`Cannot detect external IP address using STUN`))
     }
+
+    socket.on('message', listener)
   })
 }
 
-function performStunRequests(usableMultiaddrs: Multiaddr[], tIds: string[], socket: Socket): Promise<boolean>[] {
-  const result: Promise<boolean>[] = []
+function getUsableResults(results: Request[], runningLocally = false): Request[] {
+  let filtered: Request[] = []
 
-  for (const [index, ma] of usableMultiaddrs.entries()) {
-    if (![CODE_IP4, CODE_IP6, CODE_DNS4, CODE_DNS6].includes(ma.tuples()[0][0])) {
-      error(`Cannot contact STUN server ${ma.toString()} due invalid address.`)
+  for (const result of results) {
+    if (!result.response) {
       continue
     }
 
-    const nodeAddress = ma.nodeAddress()
-
-    const res = stun.createBindingRequest(tIds[index]).setFingerprintAttribute()
+    switch (result.response.family) {
+      case 'IPv6':
+        // STUN over IPv6 is not yet supported
+        break
+      case 'IPv4':
+        const u8aAddr = ipToU8aAddress(result.response.address, 'IPv4')
+        // If running locally, get only local addresses,
+        // otherwise only get public addresses
+        if ((isPrivateAddress(u8aAddr, 'IPv4') || isLocalhost(u8aAddr, 'IPv4')) == runningLocally) {
```

Comment:
If running locally this code adds **both** private and local addressed. Are you sure this is what you want?

Reply:
That's a good point, thanks :-)

Reply:
Yes, because it reflects our use cases:

Unit tests:
- run multiple instances on one machine, hence STUN will return `localhost:<somePort>`

CI tests:
- run multiple instances in *same* local network, e.g. sth like `10.x.y.z`, hence returning a private address is expected.